cc_utilities/legacy_upload.py

```diff
@@ -472,62 +477,191 @@ def get_validation_fn(col_name, data_dict):
         )
 
 
-def validate_legacy_case_data(df, data_dict):
-    """Validate user-supplied legacy case data based on a data dictionary
+def accumulate_row_validation_problems(
+    row,
+    colname=None,
+    is_missing_required_error=False,
+    fails_required_one_ofs=False,
+    required_one_ofs=None,
+):
+    """Helper for accumulating validation problems across columns for a given row
 
     Args:
-        df (object): A Pandas dataframe generated by loading user-supplied legacy
-            case data
-        data_dict (dict): A dictionary whose keys are `col_name`s and
-            whose values are a dict whose keys are field, group, allowed_values,
-            data_type, and required
+        row (object): A Pandas dataframe row
+        colname (string): Optional. A name of a column validating for.
+        is_missing_required_error (bool): Defaults to False. If `True` will generate
+            a validation error message about a missing required value
+        fails_required_one_ofs (bool): Defaults to False. If `True` will generate
+            a validation error message about a row that doesn't have at least one
+            of several "one-of" required values
+        required_one_ofs (list): A list of required one ofs to be included in a
+            validation message about required one ofs.
+
     Returns:
-        obj: A copy of the original df, with additional columns with validation data.
+        str: A string of accumulated validation problem messages (i.e., what was
+        there already + new validation problem messages)
     """
-    # helper function for accumulating validation problems across columns for a given
-    # row
-    def _accumulate_row_validation_problems(
-        row, colname, is_missing_required_error=False
-    ):
+    required_one_ofs = required_one_ofs if required_one_ofs else []
+
+    if fails_required_one_ofs:
+        assert required_one_ofs
+    assert not all([is_missing_required_error, fails_required_one_ofs])
+    if is_missing_required_error:
+        msg = f"A value must be supplied for {colname}"
+    elif fails_required_one_ofs:
         msg = (
-            f"Invalid value for {colname}"
-            if not is_missing_required_error
-            else f"A value must be supplied for {colname}"
-        )
-        return (
-            ", ".join([row["validation_problems"], msg])
-            if row["validation_problems"]
-            else msg
+            f"A valid value must be supplied for one of the following "
+            f"columns: {required_one_ofs}"
```

Comment:
Not sure this is easily fixable, but I do notice that the values in "required_one_ofs" that are shown to users here may differ from the values in the original spreadsheet, when one (or more) of the columns is internally renamed...

Reply:
Yeah, I noticed that too. That's an overall issue, even without this logic. I'll file an issue, but think we should attack that separately (if at all).

Reply:
Created [this issue](https://github.com/caktus/commcare-utilities/issues/39 )

And think it's easy enough to fix. The logic for validation reports and final reports should just get moved into separate functions that can optionally get a column mapping parameter.